### PR TITLE
User can fork recipe

### PR DESCRIPTION
**Pull request by alghazu**
### What we did:
* add two attributes to recipes `original_recipe_id` and `forked_recipes_ids`
* add `forks_controller`
* add `fork` and `forked?` methods in `Recipe` model
* add fork button in show view with condition
* add "Forked from: [original recipe]" and "Forks: [number]" in show view

* Semaphore: [Build Status](https://semaphoreci.com/craftacademy/food_hub)
* Coveralls: [Coverage Status](https://coveralls.io/github/CraftAcademy/food_hub)

### CODE REVIEW

**Review by GeorgeJLewis — APPROVED** [review]
Looks ok to me

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -32,11 +33,17 @@ def update
...
+      flash[:alert] = @recipe.errors.full_messages.first
       render :edit
     end
   end
 
+  def fork
+    @forked_recipe = @recipe.fork(current_user)
+    flash[:notice] = "Recipe was sucessfully forked"
+    redirect_to edit_recipe_path(@forked_recipe)
+  end
```
Comment (tochman) [inline app/controllers/recipes_controller.rb]: I would suggest creating a `ForksController` with a `create` action to keep the whole thing RESTful. 

I would also take this a bit further and create a `Fork` class that would inherit from the Recipe class.

File: app/models/recipe.rb [hunk 2]
```diff
@@ -12,4 +12,17 @@ def notify
...
                                   message: "<p>#{self.title} was created!</p>"
   end
+
+  def fork(user)
+    forked_recipe = self.dup
+    forked_recipe.user = user
+    forked_recipe.original_recipe_id << self.id
+    forked_recipe.title = 'Forked ' + forked_recipe.title
+    forked_recipe.forked_recipes_ids = []
+    forked_recipe.save
```
Comment (tochman) [inline app/models/recipe.rb]: I would create an `attributes` object and use it to update the new fork. (after you `.dup`'ed it
```ruby
attributes = {user: user, original_recipe_id: self.id, title: 'Forked ' + forked_recipe.title, forked_recipes_ids: []}
forked_recipe.uppdate_attributes(attributes)
```

File: app/policies/recipe_policy.rb [hunk 3]
```diff
@@ -7,4 +7,8 @@ def update?
   def edit?
     update?
   end
+  
+  def fork?
+    user.present?
```
Comment (tochman) [inline app/policies/recipe_policy.rb]: Well, the creator of the recipe should not be able to fork his own recipes, right?

File: app/views/recipes/show.html.haml [hunk 4]
```diff
@@ -1,23 +1,34 @@
 %h3= @recipe.title
-%p= @recipe.description
-%p= @recipe.ingredients
-%p= @recipe.directions
 %p
   Posted by:
   = mail_to @recipe.user.email
+%p
+  - if @recipe.original_recipe_id.any?
```
Comment (tochman) [inline app/views/recipes/show.html.haml]: I would create an instance method on the `Recipe` class called `fork?` to determine if this is a fork we are dealing with, rather then having this logic in the view.

File: config/routes.rb [hunk 5]
```diff
@@ -5,6 +5,6 @@
   resources :recipes do
     resources :comments
   end
-
+  get :fork, controller: :recipes, action: :fork
```
Comment (tochman) [inline config/routes.rb]: IMO, this should/could be a POST request

File: dump.rdb [hunk 6]
```diff
@@ -0,0 +1 @@
+REDIS0006�ܳC�Z��V
```
Comment (tochman) [inline dump.rdb]: ??

File: features/step_definitions/basic_steps.rb [hunk 7]
```diff
@@ -50,6 +50,11 @@
   visit edit_recipe_path(recipe)
 end
 
+Given("I am on the {string} page") do |recipe_title|
```
Comment (tochman) [inline features/step_definitions/basic_steps.rb]: You should consider specifically state that you want to be on the `show` or `details` page for a recipe

File: features/user_can_fork_recipe.feature [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+Scenario: User forks a recipe edits its title
+  Given I am on the "Meatball-Stew" page
+  When I click "Fork"
+  Then I should see "Recipe was sucessfully forked"
+  And I should be on the "Forked Meatball-Stew" edit page
+  When I fill in "Title" with "Maran's Meatball-Stew"
+  And I click "Update Recipe"
+  Then I should see "Maran's Meatball-Stew"
+  And I should see "Posted by: maran@test.com"
+  And I should see "Forked from: Meatball-Stew"
```
Comment (tochman) [inline features/user_can_fork_recipe.feature:27]: I would like to see a scenario for the original author of the recipe trying to fork it.

File: app/controllers/recipes_controller.rb (hunk 1 above)
Comment (alghazu) [reply app/controllers/recipes_controller.rb]: Now it is RESTful, but after consideration will skip creating a `fork` model.

File: app/models/recipe.rb (hunk 2 above)
Comment (alghazu) [reply app/models/recipe.rb]: :+1:

File: app/policies/recipe_policy.rb (hunk 3 above)
Comment (alghazu) [reply app/policies/recipe_policy.rb]: Yes he can fork his own recipes, e.g. User posted his grandma original recipe, then forks to show his own modern version of it.

File: app/views/recipes/show.html.haml (hunk 4 above)
Comment (alghazu) [reply app/views/recipes/show.html.haml]: :+1:

File: config/routes.rb (hunk 5 above)
Comment (alghazu) [reply config/routes.rb]: :+1:

File: dump.rdb (hunk 6 above)
Comment (alghazu) [reply dump.rdb]: :+1:

File: features/step_definitions/basic_steps.rb (hunk 7 above)
Comment (alghazu) [reply features/step_definitions/basic_steps.rb]: :+1:

File: features/user_can_fork_recipe.feature (hunk 8 above)
Comment (alghazu) [reply features/user_can_fork_recipe.feature:27]: Author can fork his own recipe.

File: app/controllers/recipes_controller.rb [hunk 9]
```diff
@@ -53,13 +54,13 @@ def recipe_params
...
  end
 
  def authorize_record
     authorize @recipe
  end
 
- 
-end
\ No newline at end of file
+
```
Comment (GeorgeJLewis) [inline app/controllers/recipes_controller.rb]: Delete the white space

**Review by GeorgeJLewis — CHANGES_REQUESTED** [review]
Just one comment on white space but thats it

File: app/controllers/recipes_controller.rb (hunk 9 above)
Comment (alghazu) [reply app/controllers/recipes_controller.rb]: :+1:

**Review by GeorgeJLewis — APPROVED** [review]
LGTM

**Review by tochman — APPROVED** [review]
Looks good

File: features/user_can_fork_recipe.feature (hunk 8 above)
Comment (tochman) [reply features/user_can_fork_recipe.feature:27]: Okay, thx

**Pull request merged**